manubot/cite/citeproc.py

```diff
@@ -22,29 +23,29 @@ def csl_item_passthrough(csl_item, set_id=None, prune=True):
     http://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html
     https://github.com/citation-style-language/schema/blob/master/csl-data.json
     """
+    csl_item = CSL_Item(csl_item) # wrap dictionary as CSL_Item dictionary
     if set_id is not None:
         csl_item['id'] = set_id
     logging.debug(f"Starting csl_item_passthrough with{'' if prune else 'out'} CSL pruning for id: {csl_item.get('id', 'id not specified')}")
 
     # Correct invalid CSL item types
-    # See https://github.com/CrossRef/rest-api-doc/issues/187
-    if 'type' in csl_item:
-        csl_item['type'] = csl_item_type_fixer.get(csl_item['type'], csl_item['type'])
+    csl_item = csl_item.correct_invalid_type()
 
     if prune:
         # Remove fields that violate the CSL Item JSON Schema
         csl_item, = remove_jsonschema_errors([csl_item])
```

Comment:
I think this will return a new csl_item (a deepcopy). This is probably why `assert isinstance(csl_item, CSL_Item)` sometimes failed below? Anyways this seems like a problem. Easy solution for now would be to wrap it in CSL_Item.

Reply:
> I think this will return a new dict for csl_item (a deepcopy). This is probably why `assert isinstance(csl_item, CSL_Item)` sometimes failed below? Anyways this seems like a problem. Easy solution for now would be to wrap it in CSL_Item.

Here is something I understood about the conversation - we should treat the two problems separately:

1. behavior of `CSL_Item` constructor
2. reuse of a specific CSL item in `csl_item_passthrough()` function.

The constructor more or less behaves like a usual dictionary, so adding more safequards into the constructor is struggling with a rempte problem that originates elsewhere in the program, namely - at (2).

The proposed solution is using a deepcopy to copy csl_item at start of `csl_item_passthrough()`. this way we can relate why we are doing it to what goes on in `csl_item_passthrough()`. Hiding a safeguard in a constructor imo breaks a logic why we are doing it. 

In `csl_item_passthrough()`

Reply:
copy.deepcopy returns a CSL_Item not a plain dict, so my [comment above](https://github.com/manubot/manubot/pull/150#discussion_r335805651) was mistaken. I don't think `csl_item_passthrough` needs to return a deepcopy of its input.

My view that the CSL_Item constructor should create a deep copy is unrelated to csl_item_passthrough. As I mentioned before, I would like to make csl_item_passthrough a method of the CSL_Item class after we've merge this PR. My reasoning for putting the deepcopy in the constructor is explained in https://github.com/manubot/manubot/pull/150#discussion_r335804407. Basically, I thinks it's a very rare instance where the user will actually want to have deep objects as pointers but the top-level dict as a shallow copy, yet this is likely to cause problems.

Reply:
Ok, but why is there a concern about `CSL_Item` deepcopy vs existing implementation of a `csl_item` as a dict? You are not deepcoping the dict, why deepcopy `CSL_Item`? Where does the mutable version of a dictionary exist?

The retriever functions like `DOI('blah').csl_item()` should return the CSL_Item, passthrough takes a `CSL_Item` and returns a `CSL_Item`. After refactoring `CSL_Item` is a base data structure - there should be no uses of dict in place of `CSL_Item`. 

Plugging `super().__init__(copy.deepcopy(dictionary))` to constructor is easy, let's do it, but also important to indentify the source of disturbance about inline change - maybe some other things need to change as well. 

A checklist we are on the same note:
- [ ] `Handle.csl_item()` returns `CSL_Item`
- [ ] a user does not create `CSL_Item` any objects (unless in tests) 
- [ ] worker functions like `csl_item_passthrough()` create a new  `CSL_Item` object, their behaviour is fixed.

Reply:
> why is there a concern about CSL_Item deepcopy vs existing implementation of a csl_item as a dict?

Prior to this PR, the implementation does not copy the csl_item dict after its creation (ignoring `csl_item_passthrough`). By creating a CSL_Item class, we're introducing a copying of a csl item dict. So previously, since we were just using dict and never copying it, the question of copy versus deepcopy did not arise.

> After refactoring CSL_Item is a base data structure - there should be no uses of dict in place of CSL_Item.

I agree. Once we have the CSL_Item class, we should migrate all csl-item dictionaries to use the class. The question is how we should design the CSL_Item class in terms of when it copies versus deepcopies versus edits in-place. My proposal is that the constructor should deepcopy rather than shallow copy. Furthermore, I think methods should default to editing in place, which is currently how you have it.

>  a user does not create CSL_Item any objects (unless in tests)

I think users should be able to create CSL_Items if they are manually constructing CSL JSON. Although generally I think they'll just be calling functions that return CSL_Item objects.

>  worker functions like csl_item_passthrough() create a new CSL_Item object, their behaviour is fixed.

I think we eventually want to move pruning as a method, in which case perhaps it should edit in place (unless that is not possible). As far as functions that are defined elsewhere that use CSL_Item objects, I am not sure whether they should edit in-place or return a new CSL_Item. I think that may depend on the application.

Reply:
> By creating a CSL_Item class, we're introducing a copying of a csl item dict.

You are not creating a dict anymore, you are creating a `CSL_Item`. The only place where a concern applies is how `Handle.csl_item()` should work. 

```python 
csl_item = DOI('blah').csl_item()
```

> Furthermore, I think methods should default to editing in place, which is currently how you have it.

We are changing `self` in place and assigning to a new variable. The inline methods did not work out to replicate `csl_item_passthrough`, that's why I had 00175471cdaa18c916ba433c4554c06365d1e3af

> I think users should be able to create CSL_Items if they are manually constructing CSL JSON.

Is this anywhere in the program now?

Reply:
I'll follow up with commit for a deepcopy constructor, even though my opinion it attempts to patch a problem that is likely due to behaviour of other parts of program. 

If that is a real concern, I suggest @dhimmel may want to add a unit test that demonstartes the benefits of deepcopy.

Reply:
Maybe `csl_item_` should be back as `csl_item` in `csl_item_passthrough()`